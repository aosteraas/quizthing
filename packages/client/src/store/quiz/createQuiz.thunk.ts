import { QuizActions } from './quiz.slice';
import { CreateQuizDto, QuizService } from '../../services';
import { AppThunk } from '..';

export function createQuiz(data: CreateQuizDto): AppThunk {
  return async (dispatch) => {
    dispatch(QuizActions.create());
    try {
      const payload = await QuizService.create(data);
      dispatch(QuizActions.created(payload));
    } catch (e) {
      dispatch(QuizActions.createFailed(e));
    }
  };
}
