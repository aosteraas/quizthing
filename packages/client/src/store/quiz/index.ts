export * from './createQuiz.thunk';
export * from './quiz.slice';
