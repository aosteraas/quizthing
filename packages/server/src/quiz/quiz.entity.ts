import {
  BaseEntity,
  Entity,
  PrimaryGeneratedColumn,
  Column,
  OneToMany,
} from 'typeorm';
import { Question } from 'src/question/question.entity';

@Entity()
export class Quiz extends BaseEntity {
  @PrimaryGeneratedColumn()
  id: number;

  @Column()
  title: string;

  @Column({ type: 'timestamptz' })
  created: string;

  @OneToMany(() => Question, (question) => question.question, { eager: false })
  questions: Question[];
}
