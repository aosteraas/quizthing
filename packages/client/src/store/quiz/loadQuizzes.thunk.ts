import { AppThunk } from '../index';
import { QuizActions } from './quiz.slice';
import { QuizService } from '../../services';

export function loadQuizzes(): AppThunk {
  return async (dispatch) => {
    try {
      dispatch(QuizActions.load(true));
      const res = await QuizService.load();
      dispatch(QuizActions.loaded(res));
    } catch (e) {
      console.error(e);
      dispatch(QuizActions.load(false));
    }
  };
}
