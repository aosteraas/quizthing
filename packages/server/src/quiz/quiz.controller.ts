import {
  Controller,
  Post,
  UsePipes,
  ValidationPipe,
  Body,
  Delete,
  Param,
  ParseIntPipe,
  HttpCode,
  HttpStatus,
} from '@nestjs/common';
import { QuizService } from './quiz.service';
import { GetUser } from '../auth/get-user-decorator';
import { User } from '../auth/user.entity';
import { CreateQuizDto } from './dto/create-quiz.dto';

@Controller('quiz')
export class QuizController {
  constructor(private readonly quizService: QuizService) {}

  @Post()
  @UsePipes(ValidationPipe)
  async create(@Body() createQuizDto: CreateQuizDto, @GetUser() user: User) {
    const quiz = await this.quizService.createQuiz(createQuizDto, user);
    return quiz;
  }

  @Delete('/:id')
  @HttpCode(HttpStatus.NO_CONTENT)
  async deleteQuiz(
    @Param('id', ParseIntPipe) id: number,
    @GetUser() user: User,
  ) {
    return await this.quizService.deleteQuiz(id, user);
  }
}
