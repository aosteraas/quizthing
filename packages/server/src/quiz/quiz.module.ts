import { Module } from '@nestjs/common';
import { TypeOrmModule } from '@nestjs/typeorm';
import { QuizService } from './quiz.service';
import { QuizController } from './quiz.controller';
import { QuizRepository } from './quiz.repository';

@Module({
  imports: [TypeOrmModule.forFeature([QuizRepository])],
  controllers: [QuizController],
  providers: [QuizService],
})
export class QuizModule {}
