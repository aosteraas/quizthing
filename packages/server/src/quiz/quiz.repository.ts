import { EntityRepository, Repository } from 'typeorm';
import { Quiz } from './quiz.entity';
import { CreateQuizDto } from './dto/create-quiz.dto';
import { User } from '../auth/user.entity';

@EntityRepository(Quiz)
export class QuizRepository extends Repository<Quiz> {
  async createQuiz(createQuizDto: CreateQuizDto, user: User) {
    const quiz = new Quiz();
    quiz.title = createQuizDto.title;
    quiz.description = createQuizDto.description ?? '';
    quiz.user = user;

    await quiz.save();

    delete quiz.user;

    return quiz;
  }
}
