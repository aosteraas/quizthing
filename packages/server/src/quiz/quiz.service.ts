import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { QuizRepository } from './quiz.repository';

@Injectable()
export class QuizService {
  constructor(
    @InjectRepository(QuizRepository)
    private readonly quizRepository: QuizRepository,
  ) {}

  async createQuiz() {
    // this.quizRepository.createQuiz
  }
}
