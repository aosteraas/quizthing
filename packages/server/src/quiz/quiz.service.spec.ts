import { Test, TestingModule } from '@nestjs/testing';
import { QuizService } from './quiz.service';
import { QuizRepository } from './quiz.repository';
import { User } from '../auth/user.entity';
import { Quiz } from './quiz.entity';
import { NotFoundException } from '@nestjs/common';

const mockQuizRepository = () => ({
  createQuiz: jest.fn(),
  delete: jest.fn(),
});

describe('QuizService', () => {
  let service: QuizService;
  let repository;
  const mockUser = {
    id: 1,
    username: 'testuser',
    email: 'fake@email.com',
    password: 'fake password',
    salt: 'some salt',
  } as User;
  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        QuizService,
        { provide: QuizRepository, useFactory: mockQuizRepository },
      ],
    }).compile();

    service = module.get<QuizService>(QuizService);
    repository = module.get<QuizRepository>(QuizRepository);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });
  it('createQuiz returns a new quiz without the user', async () => {
    const createQuizDto = { title: 'some quiz', desc: 'something' };
    const quizEntity = {
      ...createQuizDto,
      id: 1,
      created: new Date('2020-06-24T09:42:28.004Z'),
      questions: [],
      userId: mockUser.id,
    };
    repository.createQuiz.mockResolvedValue(quizEntity);
    const quiz = await service.createQuiz(createQuizDto, mockUser);

    expect(quiz).toMatchSnapshot();
    expect(quiz.user).toBeUndefined();
  });
  it('deleteQuiz throws not found exception when matching quiz not found', async () => {
    repository.delete.mockResolvedValue({ affected: 0 });
    expect(service.deleteQuiz(1, mockUser)).rejects.toThrow(NotFoundException);
  });
  it('deleteQuiz returns void when deleting a quiz successfully', async () => {
    repository.delete.mockResolvedValue({ affected: 1 });
    expect(service.deleteQuiz(1, mockUser)).resolves.toBeUndefined();
  });
  it('updateQuiz', () => {
    expect(() => service.updateQuiz()).toThrow();
  });
});
