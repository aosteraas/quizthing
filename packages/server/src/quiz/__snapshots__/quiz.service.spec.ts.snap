// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuizService createQuiz returns a new quiz without the user 1`] = `
Object {
  "created": 2020-06-24T09:42:28.004Z,
  "desc": "something",
  "id": 1,
  "questions": Array [],
  "title": "some quiz",
  "userId": 1,
}
`;
